File: app/schemas/product.py
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field


class ProductCreate(BaseModel):
    product_code: str = Field(min_length=1, max_length=64)
    category: str = Field(min_length=1, max_length=64)
    brand: str = Field(min_length=1, max_length=64)
    name_cn: str = Field(min_length=1, max_length=150)
    name_en: str | None = Field(default=None, max_length=150)
    specification: str | None = Field(default=None, max_length=255)
    matrix: str | None = Field(default=None, max_length=120)
    original_price: Decimal = Field(ge=0)
    unit: str | None = Field(default=None, max_length=30)
    remark: str | None = None


class ProductUpdate(ProductCreate):
    pass


class ProductRead(BaseModel):
    id: uuid.UUID
    product_code: str
    category: str
    category_display: str
    brand: str
    brand_display: str
    name_cn: str
    name_en: str | None
    specification: str | None
    matrix: str | None
    original_price: Decimal
    unit: str | None
    remark: str | None
    created_at: datetime


class ProductImportResult(BaseModel):
    imported_count: int
    skipped_count: int

File: frontend/src/services/product.ts
import { request } from '../utils/request';

export interface IProduct {
  id: string;
  product_code: string;
  category: string;
  category_display: string;
  brand: string;
  brand_display: string;
  name_cn: string;
  name_en: string | null;
  specification: string | null;
  matrix: string | null;
  original_price: number | string;
  unit: string | null;
  remark: string | null;
  created_at: string;
}

export interface ProductPayload {
  product_code: string;
  category: string;
  brand: string;
  name_cn: string;
  name_en?: string | null;
  specification?: string | null;
  matrix?: string | null;
  original_price: number;
  unit?: string | null;
  remark?: string | null;
}

export interface ImportResult {
  imported_count: number;
  skipped_count: number;
}

export async function getProducts(storeId?: string): Promise<IProduct[]> {
  const response = await request.get<IProduct[]>('/api/products', {
    params: {
      store_id: storeId || undefined,
    },
  });
  return response.data;
}

export async function createProduct(payload: ProductPayload): Promise<IProduct> {
  const response = await request.post<IProduct>('/api/products', payload);
  return response.data;
}

export async function updateProduct(productId: string, payload: ProductPayload): Promise<IProduct> {
  const response = await request.put<IProduct>(`/api/products/${productId}`, payload);
  return response.data;
}

export async function importProducts(file: File): Promise<ImportResult> {
  const formData = new FormData();
  formData.append('file', file);

  const response = await request.post<ImportResult>('/api/products/import', formData, {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
  });
  return response.data;
}

export async function exportProducts(): Promise<Blob> {
  const response = await request.get<Blob>('/api/products/export', {
    responseType: 'blob',
  });
  return response.data;
}
